server/src/main/java/org/elasticsearch/transport/Header.java

```diff
@@ -0,0 +1,118 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Set;
+
+public class Header {
+
+    private static final String RESPONSE_NAME = "NO_ACTION_NAME_FOR_RESPONSES";
+
+    private final int networkMessageSize;
+    private final Version version;
+    private final long requestId;
+    private final byte status;
+    private String actionName;
+    private Tuple<Map<String, String>, Map<String, Set<String>>> headers;
+
+    Header(int networkMessageSize, long requestId, byte status, Version version) {
+        this.networkMessageSize = networkMessageSize;
+        this.version = version;
+        this.requestId = requestId;
+        this.status = status;
+    }
+
+    public int getNetworkMessageSize() {
+        return networkMessageSize;
+    }
+
+    Version getVersion() {
+        return version;
+    }
+
+    long getRequestId() {
+        return requestId;
+    }
+
+    byte getStatus() {
+        return status;
+    }
+
+    boolean isRequest() {
+        return TransportStatus.isRequest(status);
+    }
+
+    boolean isResponse() {
+        return TransportStatus.isRequest(status) == false;
+    }
+
+    boolean isError() {
+        return TransportStatus.isError(status);
+    }
+
+    boolean isHandshake() {
+        return TransportStatus.isHandshake(status);
+    }
+
+    boolean isCompressed() {
+        return TransportStatus.isCompress(status);
+    }
+
+    public String getActionName() {
+        return actionName;
+    }
+
+    boolean needsToReadVariableHeader() {
+        return headers == null;
+    }
+
+    Tuple<Map<String, String>, Map<String, Set<String>>> getHeaders() {
+        return headers;
+    }
+
+    void finishParsingHeader(StreamInput input) throws IOException {
+        setHeaders(ThreadContext.readHeadersFromStream(input));
+
+        if (isRequest()) {
+            if (version.before(Version.V_8_0_0)) {
+                // discard features
+                input.readStringArray();
+            }
+            setActionName(input.readString());
+        } else {
+            setActionName(RESPONSE_NAME);
+        }
+    }
+
+    void setHeaders(Tuple<Map<String, String>, Map<String, Set<String>>> headers) {
+        this.headers = headers;
```

Comment:
perhaps remove method and make variable package-private, adding comment that it is directly accessed by tests